core/contextmenu_items.js

```diff
@@ -270,12 +270,230 @@ Blockly.ContextMenuItems.registerWorkspaceOptions_ = function() {
+Blockly.ContextMenuItems.registerComment = function() {
+  var commentOption = {};
+  commentOption.displayText = function(scope) {
+    if (scope.block.getCommentIcon()) {
+      // If there's already a comment,  option is to remove.
+      return Blockly.Msg['REMOVE_COMMENT'];
+    }
+    // If there's no comment yet, option is to add.
+    return Blockly.Msg['ADD_COMMENT'];
+  };
+  commentOption.preconditionFn = function(scope) {
+    var block = scope.block;
+    // IE doesn't support necessary features for comment editing.
+    if (!Blockly.utils.userAgent.IE && !block.isInFlyout && block.workspace.options.comments &&
+        !block.collapsed_ && block.isEditable()) {
+      return 'enabled';
+    }
+    return 'hidden';
+  };
+  commentOption.callback = function(scope) {
+    var block = scope.block;
+    if (block.getCommentIcon()) {
+      block.setCommentText(null);
+    } else {
+      block.setCommentText('');
+    }
+  };
+  commentOption.scopeType = Blockly.ContextMenuRegistry.ScopeType.BLOCK;
+  commentOption.id = 'blockComment';
+  commentOption.weight = 0;
+  Blockly.ContextMenuRegistry.registry.register(commentOption);
+};
+
+/** Option to inline variables. */
+Blockly.ContextMenuItems.registerInline = function() {
+  var inlineOption = {};
+  inlineOption.displayText = function(scope) {
+    return (scope.block.getInputsInline()) ?
+        Blockly.Msg['EXTERNAL_INPUTS'] : Blockly.Msg['INLINE_INPUTS'];
+  };
+  inlineOption.preconditionFn = function(scope) {
+    var block = scope.block;
+    if (!block.isInFlyout && block.isMovable() && !block.collapsed_) {
+      for (var i = 1; i < block.inputList.length; i++) {
+        // Only display this option if there are two value or dummy inputs next to each other.
+        if (block.inputList[i - 1].type != Blockly.NEXT_STATEMENT &&
+            block.inputList[i].type != Blockly.NEXT_STATEMENT) {
+          return 'enabled';
+        }
+      }
+    }
+    return 'hidden';
+  };
+  inlineOption.callback = function(scope) {
+    scope.block.setInputsInline(!scope.block.getInputsInline());
+  };
+  inlineOption.scopeType = Blockly.ContextMenuRegistry.ScopeType.BLOCK;
+  inlineOption.id = 'blockInline';
+  inlineOption.weight = 0;
+  Blockly.ContextMenuRegistry.registry.register(inlineOption);
+};
+
+/** Option to collapse or expand a block. */
+Blockly.ContextMenuItems.registerCollapseExpandBlock = function() {
+  var collapseExpandOption = {};
+  collapseExpandOption.displayText = function(scope) {
+    if (scope.block.collapsed_) {
+      return Blockly.Msg['EXPAND_BLOCK'];
+    }
+    return Blockly.Msg['COLLAPSE_BLOCK'];
+  };
+  collapseExpandOption.preconditionFn = function(scope) {
+    var block = scope.block;
+    if (!block.isInFlyout && block.isMovable()) {
+      return 'enabled';
+    }
+    return 'hidden';
+  };
+  collapseExpandOption.callback = function(scope) {
+    scope.block.setCollapsed(!scope.block.collapsed_);
```

Comment:
collapsed_ is protected, use `isCollapsed()` instead.

Reply:
Done! also found that there is a workspace.cancelCurrentGesture(). For the undo and redo stack, I added getters and made them package, but also happy to do any of the options we discussed.